### DISCUSSION TITLE
Students logging into personal github accounts from Azure Labs VMs getting "secondary rate limit exceeded" on login.

### DISCUSSION THREAD

**Discussion by JeffryGonzalez** [community/community#120740] [Code Security]
### Select Topic Area

Bug

### Body

I teach my classes using Microsoft's Azure Labs students. Each student gets a VM for the duration of the class. In the past the students were able to log into their Github accounts no problem (from the VMs). Now many of them are getting "Secondary Rate Limit Exceeded" errors. I'm assuming because the VMs all seem to originate from the same IP (they differentiate by using ports), that IP address is being rate limited. Is this a recent change? Is there something I could do to prevent this?

Thanks!

**Comment by david-wei-01001** [comment]
The "Secondary Rate Limit Exceeded" error from GitHub typically indicates that there's been an unusual amount of activity from a single IP address that resembles abusive behavior, even if the actions are legitimate. This rate limiting is an automated security feature to prevent abuse of the platform. If all the VMs in your Azure Lab are indeed sharing an IP address and students are trying to log in around the same time, this could trigger the rate limit.

Here's what you can do:

1. **Stagger Logins**: If possible, have students log in at staggered times rather than all at once to avoid hitting the rate limit.

2. **Contact GitHub Support**: Explain your situation to GitHub Support. They can provide more information on the rate limits and may be able to whitelist your IP address or provide guidance on best practices for classroom settings.

3. **Check for Changes**: It's possible that GitHub has recently changed their rate limiting policies. Review GitHub's documentation on rate limiting or check their blog and public communications for any recent updates that might affect how they handle numerous logins from the same IP.

4. **Educate Students on Best Practices**: Inform students about the importance of not over-using GitHub actions such as authentication requests, API calls, etc., from the same IP address.

5. **Use Personal Access Tokens (PATs)**: If appropriate, have students use PATs for authentication with Git operations instead of logging into GitHub multiple times. This may reduce the frequency of authentication requests from the same IP address.

6. **VPN or Proxy Services**: Using VPNs or proxy services to give each VM a unique IP address might help, but this could introduce complexity and additional costs.

7. **Azure Feedback**: Contact Azure to report this issue since it's related to their lab environment setup. Azure might be able to provide a solution such as assigning different outgoing IP addresses for the VMs.

8. **Alternate Education Tools**: Look into GitHub Classroom or GitHub Education benefits, which might provide tools and workflows more suited to classroom use that may be less likely to trigger rate limiting.

It's also important to ensure that no automated scripts or processes on the VMs are making frequent requests to GitHub, as these could contribute to hitting the rate limit. Proper scripting and use of the GitHub API (including respectful use of the rate limits) should be part of the education you provide to your students.

It's not clear whether there was a recent change in GitHub's rate limiting policy that's causing this new issue, but contacting GitHub directly is the best way to find out and to seek a resolution.

**Comment by marci543** [comment]
I can confirm your experience (reported [here](https://github.com/community/Global-Campus-Teachers/discussions/330) but no answer yet).

**Login fails in CS computer lab with 40 students, secondary rate limit reached**
> We have a university course held in a computer laboratory, where we teach GitHub for 3 groups, each with 40 students for 3x2 hours (so normally 3 times 40 login attempts).
Several students (15 out of 40) fail to log in to GitHub web page at the start of the lab with the error message that secondary rate limit is reached. We do not use any API calls and only want to log in to the GitHub web page.
(All the 40 computers are using NAT via the same IP address.)
Is there any solution for 40 students to log in to GitHub at the start of a computer lab?
> 
> In other cases, we experienced that GitHub asks for a login verification code sent in e-mail when students try to log in but fails to send the e-mail. As a workaround, we asked them to use GitHub Mobile instead of e-mail verification, but this is only a partial solution as only some of them have it installed.

**Reply by JeffryGonzalez** [reply]
Thanks - feel ya, friend. I'm sure that the rate limit based on IP was put in place in reaction to a real need (assuming that's what is going on), but this is more than just a minor inconvenience for those of us that teach. I wonder if there is a somewhat fixed range of IPs that are used for Azure Labs (Microsoft's left hand) that could be "white listed" by Github (Microsoft's "right hand")? I teach in a (BIG) corporate environment, and will leverage that as I can to see if there is something that can be done.

**Comment by sharoon858** [comment]
response #2

**Reply by sharoon2256** [reply]
done

**Reply by sharoon858** [reply]
response 3 1

**Comment by ghost** [comment]
So proxy - students are ow using proxy to acess github
